### PR TITLE
PIMD update review

### PR DESCRIPTION
**Pull request by xmwebb**

### CODE REVIEW

File: src/Integrators/IntegratorVerlet.cu [hunk 1]
```diff
@@ -438,24 +561,24 @@ void IntegratorVerlet::nve_x() {
...
+	    for (Fix *f: state->fixes) {
+	      if ( f->isThermostat && f->groupHandle == "all" ) {
+	        std::string t = "temp";
+	        temp = f->getInterpolator(t)->getCurrentVal();
+	      }
+	    }
+	    int   nPerRingPoly = state->nPerRingPoly;
+        int   nRingPoly = state->atoms.size() / nPerRingPoly;
+	    float omegaP    = (float) state->units.boltz * temp / state->units.hbar  ;
+    	SAFECALL((nve_xPIMD_cu<<<NBLOCK(state->atoms.size()), PERBLOCK, sizeof(float3) * 2 *PERBLOCK>>>(
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: remove safecall

File: src/Integrators/IntegratorVerlet.cu [hunk 2]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
-	// to free ring-polymer evolution
-	// 3. back transform to regular coordinates
-	float invP            = 1.0 / (float) nPerRingPoly;
-	float twoPiInvP       = 2.0f * M_PI * invP;
-	float invSqrtP 	      = sqrt(invP);
-	float sqrt2           = sqrt(2.0);
-	int   halfP           = nPerRingPoly / 2;	// P must be even for the following transformation!!!
+    // k = P/2+1,...,P-1; n = 1,...,P
+    for (int k = halfP+1; k < nPerRingPoly; k++) {
+        float  sinval = sin(twoPiInvP * k * n);	// sin(2*pi*k*n/P)
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: sinf

File: src/Integrators/IntegratorVerlet.cu [hunk 3]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
+    } else {
+        tbr[threadIdx.x] = xn ;
+    }
+    if (needSync)   { __syncthreads();}
+    reduceByN<float3>(tbr, nPerRingPoly, warpSize);
+    if (useThread && amRoot)     {xsNM[threadIdx.x+halfP] = tbr[threadIdx.x]*invSqrtP;}
+
+    // k = 1,...,P/2-1; n = 1,...,P
+    for (int k = 1; k < halfP; k++) {
+        float cosval = cos(twoPiInvP * k * n);	// cos(2*pi*k*n/P)
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: cosf

File: src/Integrators/IntegratorVerlet.cu [hunk 4]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
+    // 	Cnk = \sqrt(1/P)			k = 0
+    // 	Cnk = \sqrt(2/P) cos(2*pi*k*n/P)	1<= k <= P/2 -1
+    // 	Cnk = \sqrt(1/P)(-1)^n			k = P/2
+    // 	Cnk = \sqrt(2/P) sin(2*pi*k*n/P)	P/2+1<= k <= P -1
+    // 2. advance positions/velocities by full timestep according
+    // to free ring-polymer evolution
+    // 3. back transform to regular coordinates
+    float invP            = 1.0 / (float) nPerRingPoly;
+    float twoPiInvP       = 2.0f * M_PI * invP;
+    float invSqrtP 	      = sqrt(invP);
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: sqrtf

File: src/Integrators/IntegratorVerlet.cu [hunk 5]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
+    float sqrt2           = sqrt(2.0);
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: sqrtf

File: src/Integrators/IntegratorVerlet.cu [hunk 6]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
+    // 1. Transform to normal mode positions and velocities
+    // 	xNM_k = \sum_{n=1}^P x_n* Cnk
+    // 	Cnk = \sqrt(1/P)			k = 0
+    // 	Cnk = \sqrt(2/P) cos(2*pi*k*n/P)	1<= k <= P/2 -1
+    // 	Cnk = \sqrt(1/P)(-1)^n			k = P/2
+    // 	Cnk = \sqrt(2/P) sin(2*pi*k*n/P)	P/2+1<= k <= P -1
+    // 2. advance positions/velocities by full timestep according
+    // to free ring-polymer evolution
+    // 3. back transform to regular coordinates
+    float invP            = 1.0 / (float) nPerRingPoly;
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: 1.0f

File: src/Integrators/IntegratorVerlet.cu [hunk 7]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
-	  for (int n = 1; n < nPerRingPoly+1; n++) {
-	   float3 xn     = make_float3(xs[baseIdx+n-1]);
-	   float3 vn     = make_float3(vs[baseIdx+n-1]);
-	   float  cosval = cos(twoPiInvP * k * n);	// cos(2*pi*k*n/P)
-	   xsNM[k] += xn * sqrt2 * cosval;
-	   vsNM[k] += vn * sqrt2 * cosval;
-	  }
-	}
+    for (int k = 1; k < halfP; k++) {
+        float cosval = cos(twoPiInvP * k * n);	// cos(2*pi*k*n/P)
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: cosf

File: src/Integrators/IntegratorVerlet.cu [hunk 8]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
-	   vsNM[k] += vn * sqrt2 * sinval;
-	  }
-	}
-	// orthogonalize
-	for (int n = 0; n<nPerRingPoly; n++) {
-          xsNM[0] *= invSqrtP;
-          vsNM[0] *= invSqrtP;
-	}
+    for (int k = halfP+1; k < nPerRingPoly; k++) {
+	    float  sinval = sin(twoPiInvP * k * n);	// sin(2*pi*k*n/P)
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: sinf

File: src/Integrators/IntegratorVerlet.cu [hunk 9]
```diff
@@ -195,143 +252,209 @@ __global__ void preForce_cu(int nAtoms, float4 *xs, float4 *vs, float4 *fs,
...
+	    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+        // here each bead will handle the evolution of a particular normal-mode coordinate
+	    // xk(t+dt) = xk(t)*cos(om_k*dt) + vk(t)*sin(om_k*dt)/om_k
+	    // vk(t+dt) = vk(t)*cos(om_k*dt) - xk(t)*sin(om_k*dt)*om_k
+	    // k = 0
+        if (amRoot) {
+            xsNM[threadIdx.x] += vsNM[threadIdx.x] * dt; 
+        } else {
+	        float omegaK = 2.0f * omegaP * sin( beadIdx * twoPiInvP * 0.5);
+	        float cosdt  = cos(omegaK * dt);
```
Comment (dreid1991) [inline src/Integrators/IntegratorVerlet.cu]: cosf,sinf

**Review by dreid1991 — CHANGES_REQUESTED** [review]
need to remove safecalls, convert sin/cos things to sinf, cosf, add f after a few float literals

**Comment by xmwebb** [comment]
done

**Pull request merged**